gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/CommitActivityImpl.java

```diff
@@ -0,0 +1,176 @@
+          new IOException(e),
+          null
+      );
+    }
+  }
+
+  /**
+   * Commit task states to the dataset state store.
+   * @param jobState
+   * @param taskStates
+   * @param jobContext
+   * @throws IOException
+   */
+  void commitTaskStates(State jobState, Collection<TaskState> taskStates, JobContext jobContext) throws IOException {
+    Map<String, JobState.DatasetState> datasetStatesByUrns = createDatasetStatesByUrns(taskStates);
+    final boolean shouldCommitDataInJob = JobContext.shouldCommitDataInJob(jobState);
+    final DeliverySemantics deliverySemantics = DeliverySemantics.AT_LEAST_ONCE;
+    //TODO: Make this configurable
+    final int numCommitThreads = DEFAULT_NUM_COMMIT_THREADS;
+    if (!shouldCommitDataInJob) {
+      log.info("Job will not commit data since data are committed by tasks.");
+    }
+
+    try {
+      if (!datasetStatesByUrns.isEmpty()) {
+        log.info("Persisting {} dataset urns.", datasetStatesByUrns.size());
+      }
+
+      List<Either<Void, ExecutionException>> result = new IteratorExecutor<>(Iterables
+          .transform(datasetStatesByUrns.entrySet(),
+              new Function<Map.Entry<String, JobState.DatasetState>, Callable<Void>>() {
+                @Nullable
+                @Override
+                public Callable<Void> apply(final Map.Entry<String, JobState.DatasetState> entry) {
+                  return new SafeDatasetCommit(shouldCommitDataInJob, false, deliverySemantics, entry.getKey(),
+                      entry.getValue(), false, jobContext);
+                }
+              }).iterator(), numCommitThreads,
+          ExecutorsUtils.newThreadFactory(Optional.of(log), Optional.of("Commit-thread-%d")))
+          .executeAndGetResults();
+
+      IteratorExecutor.logFailures(result, null, 10);
+
+      if (!IteratorExecutor.verifyAllSuccessful(result)) {
+        // TODO: propagate cause of failure
+        String jobName = jobState.getProperties().getProperty(ConfigurationKeys.JOB_NAME_KEY, "<job_name_stub>");
+        throw new IOException("Failed to commit dataset state for some dataset(s) of job " + jobName);
+      }
+    } catch (InterruptedException exc) {
+      throw new IOException(exc);
+    }
+  }
+
+  /**
+   * Organize task states by dataset urns.
+   * @param taskStates
+   * @return
+   */
+  public static Map<String, JobState.DatasetState> createDatasetStatesByUrns(Collection<TaskState> taskStates) {
+    Map<String, JobState.DatasetState> datasetStatesByUrns = Maps.newHashMap();
+
+    //TODO: handle skipped tasks?
+    for (TaskState taskState : taskStates) {
+      String datasetUrn = createDatasetUrn(datasetStatesByUrns, taskState);
+      datasetStatesByUrns.get(datasetUrn).incrementTaskCount();
+      datasetStatesByUrns.get(datasetUrn).addTaskState(taskState);
+    }
+
+    return datasetStatesByUrns;
+  }
+
+  private static String createDatasetUrn(Map<String, JobState.DatasetState> datasetStatesByUrns, TaskState taskState) {
```

Comment:
let's add a short java doc

Reply:
when do we not have a datasetUrn already set and when do we?

Reply:
very reasonable Q, which may deserve a comment.

on the subject of javadoc: as much as I ABSOLUTELY LOVE javadoc, I don't encourage we begin adopting it for private methods.

again, not to steer away from descriptive comments--add those early and often!--more to dissuade us away from beginning to *expect* `private` to come w/ javadoc